File: crates/plugin-runtime/extensions/logging.js
function argsToMessage(args) {
	return args.map((arg) => JSON.stringify(arg)).join(" ");
}


globalThis.console = {
	log: (...args) => {
		Deno.core.ops.op_trace(`[out]: ${argsToMessage(args)}\n`);
	},
	warn: (...args) => {
		Deno.core.ops.op_warn(`[warn]: ${argsToMessage(args)}\n`);
	},
	debug: (...args) => {
		Deno.core.ops.op_debug(`[debug]: ${argsToMessage(args)}\n`);
	},
	error: (...args) => {
		Deno.core.ops.op_error(`[error]: ${argsToMessage(args)}\n`);
	},
	info: (...args) => {
		Deno.core.ops.op_info(`[info]: ${argsToMessage(args)}\n`);
	},
}

File: crates/plugin-runtime/src/plugin_functions/logging.rs
use deno_core::{extension, op2};

#[op2(fast)]
fn op_trace(#[string] msg: &str) {
    tracing::trace!(msg);
}

#[op2(fast)]
fn op_info(#[string] msg: &str) {
    tracing::info!(msg);
}

#[op2(fast)]
fn op_debug(#[string] msg: &str) {
    tracing::debug!(msg);
}

#[op2(fast)]
fn op_warn(#[string] msg: &str) {
    tracing::warn!(msg);
}

#[op2(fast)]
fn op_error(#[string] msg: &str) {
    tracing::error!(msg);
}

extension!(
    logging,
    ops = [op_trace, op_debug, op_error, op_info, op_warn],
    esm_entry_point = "ext:logging/logging.js",
    esm = [ dir "extensions", "logging.js" ],
);
